File: querier/internal/project/project.go
package project

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
)

var ErrNotFound = errors.New("no project index")

type Coverage struct {
	JarsTotal   int      `json:"jars_total"`
	JarsIndexed int      `json:"jars_indexed"`
	JarsMissing []string `json:"jars_missing"`
}

type Pointer struct {
	SchemaVersion        int      `json:"schema_version"`
	ExtractorVersion     int      `json:"extractor_version"`
	ClasspathFingerprint string   `json:"classpath_fingerprint"`
	SessionDBPath        string   `json:"session_db_path"`
	BuiltAt              int64    `json:"built_at"`
	Coverage             Coverage `json:"coverage"`
}

func Locate(projectDir string) (string, error) {
	if projectDir != "" {
		absolute, err := filepath.Abs(projectDir)
		if err != nil {
			return "", fmt.Errorf("resolve project directory: %w", err)
		}
		candidate := filepath.Join(absolute, ".research4jar", "project.json")
		if isRegularFile(candidate) {
			return candidate, nil
		}
		return "", ErrNotFound
	}

	current, err := os.Getwd()
	if err != nil {
		return "", fmt.Errorf("get working directory: %w", err)
	}
	for {
		candidate := filepath.Join(current, ".research4jar", "project.json")
		if isRegularFile(candidate) {
			return candidate, nil
		}
		parent := filepath.Dir(current)
		if parent == current {
			return "", ErrNotFound
		}
		current = parent
	}
}

// Root returns the project root that owns .research4jar/project.json.
func Root(projectDir string) (string, error) {
	projectPath, err := Locate(projectDir)
	if err != nil {
		return "", err
	}
	return filepath.Dir(filepath.Dir(projectPath)), nil
}

func Load(path string) (Pointer, error) {
	file, err := os.Open(path)
	if err != nil {
		return Pointer{}, fmt.Errorf("open project index: %w", err)
	}
	defer file.Close()

	var pointer Pointer
	if err := json.NewDecoder(file).Decode(&pointer); err != nil {
		return Pointer{}, fmt.Errorf("decode project index: %w", err)
	}
	if pointer.SchemaVersion != 1 && pointer.SchemaVersion != 2 {
		return Pointer{}, fmt.Errorf("unsupported project schema version %d", pointer.SchemaVersion)
	}
	if pointer.SessionDBPath == "" {
		return Pointer{}, errors.New("project index has no session_db_path")
	}
	return pointer, nil
}

func isRegularFile(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.Mode().IsRegular()
}

File: core/src/main/kotlin/dev/research4jar/query/Project.kt
package dev.research4jar.query

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

/** Raised when no .research4jar/project.json is reachable. */
class ProjectNotFoundException : RuntimeException("no project index")

@JsonIgnoreProperties(ignoreUnknown = true)
data class PointerCoverage(
    @JsonProperty("jars_total") val jarsTotal: Int = 0,
    @JsonProperty("jars_indexed") val jarsIndexed: Int = 0,
    @JsonProperty("jars_missing") val jarsMissing: List<String>? = null,
)

@JsonIgnoreProperties(ignoreUnknown = true)
data class ProjectPointerData(
    @JsonProperty("schema_version") val schemaVersion: Int = 0,
    @JsonProperty("extractor_version") val extractorVersion: Int = 0,
    @JsonProperty("classpath_fingerprint") val classpathFingerprint: String = "",
    @JsonProperty("session_db_path") val sessionDbPath: String = "",
    @JsonProperty("built_at") val builtAt: Long = 0,
    @JsonProperty("coverage") val coverage: PointerCoverage = PointerCoverage(),
)

/**
 * Locates and loads .research4jar/project.json, mirroring the Go
 * querier/internal/project package: an explicit projectDir is checked
 * directly; otherwise the search walks up from the working directory.
 */
object ProjectIndex {
    private val mapper: ObjectMapper = jacksonObjectMapper()

    fun locate(projectDir: String?): Path {
        if (!projectDir.isNullOrEmpty()) {
            val candidate = Paths.get(projectDir).toAbsolutePath().normalize()
                .resolve(".research4jar").resolve("project.json")
            if (Files.isRegularFile(candidate)) return candidate
            throw ProjectNotFoundException()
        }
        var current = Paths.get("").toAbsolutePath()
        while (true) {
            val candidate = current.resolve(".research4jar").resolve("project.json")
            if (Files.isRegularFile(candidate)) return candidate
            current = current.parent ?: throw ProjectNotFoundException()
        }
    }

    /** The project root that owns .research4jar/project.json. */
    fun root(projectDir: String?): Path = locate(projectDir).parent.parent

    fun load(path: Path): ProjectPointerData {
        val pointer = try {
            mapper.readValue(Files.newInputStream(path), ProjectPointerData::class.java)
        } catch (exception: java.io.IOException) {
            throw IllegalStateException("decode project index: ${exception.message}", exception)
        }
        require(pointer.schemaVersion == 1 || pointer.schemaVersion == 2) {
            "unsupported project schema version ${pointer.schemaVersion}"
        }
        check(pointer.sessionDbPath.isNotEmpty()) { "project index has no session_db_path" }
        return pointer
    }

    /** Locate + load + manifest inference, mirroring query.ResolveProject. */
    fun resolve(projectDir: String?, home: String?): Pair<ProjectPointerData, String> {
        val pointer = load(locate(projectDir))
        var manifestPath = inferManifestPath(pointer.sessionDbPath)
        if (!home.isNullOrEmpty()) {
            manifestPath =
                dev.research4jar.indexer.Research4JarPaths.resolve(home).manifest.toString()
        }
        return pointer to manifestPath
    }

    fun inferManifestPath(sessionDbPath: String): String {
        val sessionsDir = Paths.get(sessionDbPath).parent ?: return ""
        if (sessionsDir.fileName?.toString() == "sessions") {
            return sessionsDir.parent.resolve("manifest.db").toString()
        }
        return ""
    }
}
